Starlark: optimze list(list)

stepancheg opened the pull request:
`toArray()` is more efficient than `toArray(new Object[size()])`,
so switch to `toArray()` in `Starlark.toArray(..)`.

```
A: N=23 r=8.124+-0.325 se=0.069
B: N=23 r=6.613+-0.391 se=0.083
B/A: 0.814
```

for a benchmark

```
l = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
for i in range(50000000):
    list(l)
```

Or for bundled benchmark:

```
benchmark                   ops     cpu/op    wall/op   steps/op   alloc/op
# before
bench_list_list        67108863      209ns      208ns          4       255B
# after
bench_list_list        67108863      174ns      173ns          4       223B
```

alandonovan commented on src/main/java/net/starlark/java/eval/Starlark.java:
```diff
@@ -244,7 +244,7 @@ public static void checkMutable(Mutability.Freezable x) throws EvalException {
   public static Object[] toArray(Object x) throws EvalException {
     // Specialize Sequence and Dict to avoid allocation and/or indirection.
     if (x instanceof Sequence) {
-      return ((Sequence<?>) x).toArray(new Object[((Sequence) x).size()]);
```
This is not safe: Sequence may have arbitrary subclasses and the contract for List.toArray does not mandate that the resulting class is exactly Object[], yet that is a postcondition of Starlark.toArray and it is relied upon by StarlarkList (so that later setElement calls don't fail due to ArrayStoreException).

We should add a comment here explaining that.

alandonovan commented:
> toArray() is more efficient than toArray(new Object[size()])

Perhaps StarlarkList and Tuple should override toArray(Object[]).

stepancheg commented:
* added `toArray(Object[])` specialisations in `Tuple` and `StarlarkList` (unused, but added tests)
* added `Sequence.toArray()` override to clarify it must not return `Object[]` subclasses

Note [Collection.toArray() javadoc](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Collection.html#toArray()) is self-contradictory or at least confusing:

> The returned array will be "safe" in that no references to it are maintained by this collection. (In other words, this method must allocate a new array even if this collection is backed by an array). The caller is thus free to modify the returned array.

It is safe to modify returned array if returned array is empty. In Starlark implementations of `toArray()` we return a preallocated empty instance from empty collections. But javadoc says the implementation *must* allocate a new array.

Perhaps, if we want to be strict about following the spec, we should add another `Sequence` operation, like `Sequence.starlarkToArray()` with semantics we need.

stepancheg commented:
Reverted implementation of `Starlark.toArray()`, kept only specialization to `toArray(Object[])` in `Tuple` and `StarlarkList`.

alandonovan commented:
(cr/344799998)
